Avoid unnecesary WP enqueuing and attachment race-condition

metas-ts opened the pull request.

metas-ts committed:
Avoid creation of duplicated CreateMissingICs-Workpackages

- Introduced `setCreateMissingProcessInProgress` and `isCreateMissingProcessInProgress` methods in `IInvoiceCandBL`.
- Updated `CreateMissingInvoiceCandidatesWorkpackageProcessor` and `ILHandlerModelInterceptor` to respect the new state when creating missing invoice candidates.
- Enhanced `CoalesceUtil` usage for `dateEffective` initialization in `OrderCandidatesRestController`.
- Misc. improvements in comments and formatting.

metas-ts committed:
AttachmentEntry: avoid race-condition when concurrently adding attachments

Before this, if the same attachment was loaded concurrently and the different record-references added to each, then only the refernces of the last one to be saved would prevail.

- Adjusted repository logic to utilize the new linked records structure.
- Updated `AttachmentEntryServiceTest` to align with the new implementation approach.

teosarca commented on backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/attachments/AttachmentEntry.java:
```diff
@@ -39,13 +42,27 @@ public class AttachmentEntry
+	 * Record-References that were added since the instance was loaded. Will be synched to DB on the next repo-save.
+	 */
+	@Getter(AccessLevel.PACKAGE)
+	Set<TableRecordReference> linkedRecordsToAdd = new HashSet<>();
```
use HashSet type to make it obvious that is editable.

remove `@Value` class annotation because the class is no longer immutable and 
* use EualsAndHashcode, ToStrng ann.
* use Getter
* don't like mutable fields

teosarca replied:
but tbh i still don't get it why we need it

teosarca commented on backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/attachments/AttachmentEntryRepository.java:
```diff
@@ -240,18 +237,16 @@ private void syncLinkedRecords(
 						attachmentMultiRefRecord.getRecord_ID());
-				final boolean recordReferenceWasIncluded = attachmentReferences.remove(recordReference);
-				if (!recordReferenceWasIncluded)
+				if (attachmentEntry.getLinkedRecordsToRemove().contains(recordReference))
```
why exactly it cannot be synched without having these getLinkedRecordsToRemove and getLinkedRecordsToAdd?

teosarca commented on backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/attachments/AttachmentEntryService.java:
```diff
@@ -268,7 +268,7 @@ public AttachmentEntry handleAttachmentLinks(@NonNull final AttachmentLinksReque
 		if (!Check.isEmpty(attachmentLinksRequest.getLinksToAdd()))
 		{
-			updatedEntry = updatedEntry.withAdditionalLinkedRecords(attachmentLinksRequest.getLinksToAdd());
+			updatedEntry.addLinkedRecords(attachmentLinksRequest.getLinksToAdd());
```
either u avoid these addLinkedRecords methods which add unecessary complexity
either u are returning void

teosarca commented on backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/api/IInvoiceCandBL.java:
```diff
@@ -215,10 +215,20 @@ interface IInvoiceGenerateResult
 	IAutoCloseable setUpdateProcessInProgress();
 
+	/**
+	 * See the usage of this method as well as the usage of {@link #isCreateMissingProcessInProgress(Object)}.
```
i understand nothing from this javadoc

teosarca commented on backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/api/IInvoiceCandBL.java:
```diff
@@ -215,10 +215,20 @@ interface IInvoiceGenerateResult
+	IAutoCloseable setCreateMissingProcessInProgress(@NonNull Object model);
+
+	/**
+	 * See the usage of this method as well as the usage of {@link #setCreateMissingProcessInProgress(Object)}.
```
i understand nothing from this javadoc

teosarca commented on backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/modelvalidator/ilhandler/ILHandlerModelInterceptor.java:
```diff
@@ -106,10 +107,16 @@ public void onModelChange(@NonNull final Object model, @NonNull final ModelChang
 	private void createMissingInvoiceCandidates(@NonNull final Object model)
 	{
+		// the current model might be updated **as part of** the creation of missing ICs. In that case, never enqueue a new WP
+		if(!Services.get(IInvoiceCandBL.class).isCreateMissingProcessInProgress(model))
```
make it a field

metas-ts committed:
Add AttachmentReference as root entity and refactor code accordingly

- Introduced `AttachmentReferenceRepositoryTest` covering various scenarios including save, delete, and retrieval methods.
- Refactored `TableAttachmentListenerService` and other services for improved handling of `AttachmentReference`.
- Enhanced `AttachmentMigrationService` and `StoreAttachmentService` to align with updated repository logic.
- Addressed miscellaneous formatting and dependency injection improvements.

metas-ts committed:
Merge remote-tracking branch 'origin/mergify/copy/big_coconut_uat/pr-21752' into big_coconut_uat_healtcare_racecondition

metas-ts committed:
Refactor AttachmentEntryService to AttachmentService

- Renamed `AttachmentEntryService` to `AttachmentService` across all impacted files.
- Updated method and variable references to reflect the new service name.
- Adjusted imports, annotations, and comments to match the renaming.

metas-ts committed:
Refactor imports across attachment-related classes

- Optimized and reorganized imports for better readability and consistency.
- Enhanced null-safety with `@NonNull` annotations in selected methods.
- Minor improvements in code style and formatting.

metas-ts committed:
Make `description` in `AttachmentEntryDataResource` nullable and adjust tests accordingly

metas-ts committed:
Merge branch 'big_coconut_uat' into big_coconut_uat_healtcare_racecondition

teosarca commented on backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/attachments/AttachmentEntryDataResource.java:
```diff
@@ -82,16 +81,18 @@ public boolean equals(final Object other)
 	public String getDescription()
 	{
-		return description;
+		return CoalesceUtil.coalesceNotNull(description, "");
```
why returning empty if null?... IMHO this shall be the responsability of the caller.

teosarca commented on backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/attachments/AttachmentReference.java:
```diff
@@ -0,0 +1,52 @@
+{
+	@With
+	@Nullable
+	AttachmentReferenceId id;
```
kind request: pls keep all these annotations on the same line as the row

teosarca commented on backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/attachments/AttachmentReferenceId.java:
```diff
@@ -0,0 +1,62 @@
+
+	private AttachmentReferenceId(final int repoId)
+	{
+		this.repoId = Check.assumeGreaterThanZero(repoId, "repoId");
```
consider using the actual ID column name instead o "repoId"

teosarca commented on backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/attachments/AttachmentReferenceRepository.java:
```diff
@@ -0,0 +1,184 @@
+	{
+
+		final IQueryBuilder<I_AD_Attachment_MultiRef> queryBuilder = queryBL.createQueryBuilder(I_AD_Attachment_MultiRef.class)
+				.setJoinOr();
```
calling setJoinOr on the query builder feels like a trap for the future.

teosarca commented on backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/attachments/AttachmentReferenceRepository.java:
```diff
@@ -0,0 +1,184 @@
+				.collect(ImmutableList.toImmutableList());
+	}
+
+	private AttachmentReference forRecord(@NonNull final I_AD_Attachment_MultiRef multiRefRecord)
```
static

teosarca commented on backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/attachments/AttachmentReferenceRepository.java:
```diff
@@ -0,0 +1,184 @@
+
+	public ImmutableList<AttachmentReference> saveAll(@NonNull final Collection<AttachmentReference> attachmentReferences)
+	{
+		return attachmentReferences.stream()
```
avoid SQL N+1 problem

consider loading all records in one query and then save them

teosarca commented on backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/attachments/AttachmentReferenceRepository.java:
```diff
@@ -0,0 +1,184 @@
+				.collect(ImmutableList.toImmutableList());
+	}
+
+	public AttachmentReference save(@NonNull final AttachmentReference attachmentReference)
```
for the impl o this method IMHO u shall just call saveAll(ImmutableList.o(attahcmentReference));

metas-ts replied:
i think while your reviewed yesterday, i made changes. now i think the comment is outdated

teosarca commented on backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/attachments/AttachmentService.java:
```diff
@@ -182,156 +171,216 @@ private AttachmentEntry createNewAttachmentLinkAndLinkToAllReferencedRecords(
-		if (attachmentEntriesToSave.isEmpty())
+		return Multimaps.index(
+				attachmentReferences,
+				attachmentReference -> attachmentEntryRepository.getById(attachmentReference.getAttachmentEntryId()));
```
avoid SQL N+1

teosarca commented on backend/de.metas.adempiere.adempiere/base/src/test/java/de/metas/attachments/AttachmentEntryTest.java:
```diff
@@ -0,0 +1,68 @@
+import static org.adempiere.model.InterfaceWrapperHelper.saveRecord;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class AttachmentEntryTest
```
what exactly are we testing here?

teosarca commented on backend/de.metas.adempiere.adempiere/client/src/main/java-legacy/de/schaeffer/compiere/tools/AttachmentDnDTransferHandler.java:
```diff
@@ -114,7 +114,7 @@ public boolean importData(TransferHandler.TransferSupport support)
 		final TableRecordReference tableRecordReference =TableRecordReference.of(getTableId(), getRecordId());
 
-		final AttachmentEntryService attachmentEntryService = Adempiere.getBean(AttachmentEntryService.class);
+		final AttachmentService attachmentService = Adempiere.getBean(AttachmentService.class);
```
IMHO u can consider dropping AttachmentDnDTransferHandler class

teosarca commented on backend/de.metas.adempiere.adempiere/client/src/main/java-legacy/org/compiere/apps/Attachment.java:
```diff
@@ -213,10 +213,10 @@ public void dispose()
 	private void loadAttachments()
 	{
-		final AttachmentEntryService attachmentEntryService = SpringContextHolder.instance.getBean(AttachmentEntryService.class);
+		final AttachmentService attachmentService = SpringContextHolder.instance.getBean(AttachmentService.class);
```
IMHO u can consider dropping Attachment class and code around.
on the method that opens the attachment dialog u can throw unsupported operation exception.

teosarca commented on backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/v1/order/sales/SalesOrderRestController.java:
```diff
@@ -80,11 +80,11 @@
+	private final AttachmentService attachmentService;
 
-	public SalesOrderRestController(@NonNull final AttachmentEntryService attachmentEntryService)
+	public SalesOrderRestController(@NonNull final AttachmentService attachmentService)
```
in such cases i would like to 
* drop the ctor
* annotate with RequiredArgsConstrutor
* annotate the field with NonNull

teosarca commented on backend/de.metas.document.archive/de.metas.document.archive.base/src/main/java/de/metas/document/archive/spi/impl/ReferenceableDocOutboundLogsProvider.java:
```diff
@@ -44,6 +45,8 @@ public ReferenceableDocOutboundLogsProvider()
-		return attachmentEntry.getTags().hasTagSetToTrue(AttachmentTags.TAGNAME_IS_DOCUMENT);
+		return attachmentEntry
+				.getTags()
+				.hasTagSetToTrue(AttachmentTags.TAGNAME_IS_DOCUMENT);
```
for the future, i would suggest to have hasTagSetToTrue method on attachmentEntry (The Law of Demeter / the principle of least knowledge.)

teosarca reviewed (changes requested):
check my comments

metas-ts committed:
Refactor attachment-related classes for improved immutability and caching

- Updated `loadByIds` methods to return `ImmutableList` for better immutability guarantees.
- Introduced caching in `AttachmentReferenceRepository` for efficient retrieval of referenced records.
- Replaced `AttachmentEntryId.getRepoId()` with `AttachmentEntryId.toRepoId()` for enhanced clarity.
- Streamlined and optimized attachment multi-reference retrieval logic.
- Minor improvements in naming consistency and code formatting across affected classes.

metas-ts committed:
Refactor `AttachmentService`, `AttachmentRepository`,to avoid some n+1 problems

Also Delete legacy `AttachmentDnDTransferHandler` implementation and associated references.

metas-ts committed:
Remove `AttachmentEntryTest` as three were no more tests in it

metas-ts committed:
Standardize `.gitignore` formatting by removing trailing blank line

metas-ts committed:
Fix bug with a wrong `!`

metas-ts committed:
Merge branch 'big_coconut_uat' into big_coconut_uat_healtcare_racecondition
